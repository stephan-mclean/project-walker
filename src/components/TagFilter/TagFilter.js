import React, { Component } from "react";
import PropTypes from "prop-types";
import Card from "../Card/Card";
import Button from "../Button/Button";
import InlineTagList from "../InlineTagList/InlineTagList";

class TagFilter extends Component {
  static propTypes = {
    tags: PropTypes.arrayOf(
      PropTypes.shape({
        tag: PropTypes.string,
        text: PropTypes.string,
        selected: PropTypes.bool
      })
    ).isRequired,
    onSelectionChange: PropTypes.func.isRequired
  };

  toggleAll = () => {
    const { tags, onSelectionChange } = this.props;
    const selectionValue = !tags.find(tag => tag.selected);

    let newTags = [...tags];
    newTags = newTags.map(tag => {
      tag.selected = selectionValue;
      return tag;
    });

    onSelectionChange(newTags);
  };

  toggleOne = tag => {
    const { tags, onSelectionChange } = this.props;
    let newTags = [...tags];
    newTags = newTags.map(t => {
      if (t.tag === tag.tag) {
        t.selected = !t.selected;
      }

      return t;
    });

    onSelectionChange(newTags);
  };

  areAllSelected = () => {
    const { tags } = this.props;
    return !tags.find(tag => !tag.selected);
  };

  renderTags = () => {
    const { tags } = this.props;
    return (
      <InlineTagList
        tags={tags}
        renderTagBy={(tag, index) => {
          const coloringClass = tag.selected
            ? `tag-${tag.tag}-coloring`
            : `tag-${tag.tag}-inverted-coloring`;

          let radiusClass = "no-radius";
          if (index === 0) {
            radiusClass = "radius-top-left";
          } else if (index === tags.length - 1) {
            radiusClass = "radius-bottom-right";
          }
          const onTagClick = this.toggleOne.bind(this, tag);
          return (
            <Button
              className={`btn ${radiusClass} ${coloringClass}`}
              onClick={onTagClick}
              key={`tag-filter-${tag.tag}-${index}`}
              data-testid={`tag-btn-${tag.tag}`}
            >
              {tag.text}
            </Button>
          );
        }}
      />
    );
  };

  render() {
    const { tags } = this.props;
    const allBtnClass = this.areAllSelected()
      ? "btn margin-right"
      : "btn btn--inverted margin-right";
    return (
      <Card title="Tags">
        <div className="flex">
          <div>
            <Button
              className={allBtnClass}
              onClick={this.toggleAll}
              data-testid="allBtn"
            >
              All
            </Button>
          </div>

          {this.renderTags()}
        </div>
      </Card>
    );
  }
}

export default TagFilter;
