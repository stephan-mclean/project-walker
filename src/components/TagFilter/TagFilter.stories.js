import React from "react";
import { storiesOf } from "@storybook/react";
import { State, Store } from "@sambego/storybook-state";

import TagFilter from "./TagFilter";

const store = new Store({
  tags: [
    { tag: "one", text: "one", selected: false },
    { tag: "two", text: "two", selected: false },
    { tag: "three", text: "three", selected: false },
    { tag: "four", text: "four", selected: false },
    { tag: "five", text: "five", selected: false },
    { tag: "six", text: "six", selected: false },
    { tag: "seven", text: "seven", selected: false },
    { tag: "eight", text: "eight", selected: false }
  ]
});

storiesOf("Components", module).add("TagFilter", () => {
  return (
    <State store={store}>
      {state => [
        <TagFilter
          tags={state.tags}
          onSelectionChange={newTags => store.set({ tags: newTags })}
        />
      ]}
    </State>
  );
});
