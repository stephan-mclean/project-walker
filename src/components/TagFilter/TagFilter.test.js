import React from "react";
import { mount } from "enzyme";
import TagFilter from "./TagFilter";

const emptyFn = jest.fn();

describe("Testing the TagFilter component", () => {
  it("Should render the correct buttons", () => {
    const testTags = [{ tag: "one", text: "one", selected: false }];

    const filter = mount(
      <TagFilter tags={testTags} onSelectionChange={emptyFn} />
    );

    expect(filter).toMatchSnapshot();

    const allBtn = filter.find('button[data-testid="allBtn"]');
    expect(allBtn.exists()).toBeTruthy();
    expect(allBtn.text()).toEqual("All");

    const tagBtn = filter.find('button[data-testid="tag-btn-one"]');
    expect(tagBtn.exists()).toBeTruthy();
    expect(tagBtn.text()).toEqual("one");
  });

  it("Should toggle all selection values", () => {
    const unselectedTags = [
      { tag: "one", text: "one", selected: false },
      { tag: "two", text: "two", selected: false }
    ];
    const selectedTags = [
      { tag: "one", text: "one", selected: true },
      { tag: "two", text: "two", selected: true }
    ];

    const filter = mount(
      <TagFilter tags={unselectedTags} onSelectionChange={emptyFn} />
    );

    expect(filter).toMatchSnapshot();

    const allBtn = filter.find('button[data-testid="allBtn"]');
    allBtn.simulate("click");
    expect(emptyFn).toHaveBeenCalledWith(selectedTags);
    expect(filter).toMatchSnapshot();

    allBtn.simulate("click");
    expect(emptyFn).toHaveBeenCalledWith(unselectedTags);
  });

  it("Should toggle individual selection values", () => {
    const unselectedTags = [
      { tag: "one", text: "one", selected: false },
      { tag: "two", text: "two", selected: false }
    ];

    const oneSelected = [
      { tag: "one", text: "one", selected: true },
      { tag: "two", text: "two", selected: false }
    ];

    const bothSelected = [
      { tag: "one", text: "one", selected: true },
      { tag: "two", text: "two", selected: true }
    ];

    const filter = mount(
      <TagFilter tags={unselectedTags} onSelectionChange={emptyFn} />
    );

    const tagOneBtn = filter.find('button[data-testid="tag-btn-one"]');
    tagOneBtn.simulate("click");

    expect(emptyFn).toHaveBeenCalledWith(oneSelected);

    const tagTwoBtn = filter.find('button[data-testid="tag-btn-two"]');
    tagTwoBtn.simulate("click");

    expect(emptyFn).toHaveBeenCalledWith(bothSelected);
  });

  it("Should know if all tags are selected", () => {
    const testTags = [{ tag: "one", text: "one", selected: false }];

    const filter = mount(
      <TagFilter tags={testTags} onSelectionChange={emptyFn} />
    );

    const instance = filter.instance();
    expect(instance.areAllSelected()).toBeFalsy();

    const allSelectedTestTags = [{ tag: "one", text: "one", selected: true }];

    const selectedFilter = mount(
      <TagFilter tags={allSelectedTestTags} onSelectionChange={emptyFn} />
    );

    const selectedInstance = selectedFilter.instance();
    expect(selectedInstance.areAllSelected()).toBeTruthy();
  });
});
