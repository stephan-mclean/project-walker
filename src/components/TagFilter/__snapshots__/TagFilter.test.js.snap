// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the TagFilter component Should render the correct buttons 1`] = `
<TagFilter
  onSelectionChange={[MockFunction]}
  tags={
    Array [
      Object {
        "selected": false,
        "tag": "one",
        "text": "one",
      },
    ]
  }
>
  <Card
    title="Tags"
  >
    <div
      className="container high-contrast-coloring margin-bottom radius "
    >
      <div
        className="flex flex--column"
        data-testid="titleDescContainer"
      >
        <h6>
          Tags
        </h6>
      </div>
      <div
        className="flex"
      >
        <div>
          <Button
            className="btn btn--inverted margin-right"
            data-testid="allBtn"
            onClick={[Function]}
            type="button"
          >
            <button
              className="btn btn--inverted margin-right"
              data-testid="allBtn"
              onClick={[Function]}
              type="button"
            >
              All
            </button>
          </Button>
        </div>
        <InlineTagList
          onTagClick={[Function]}
          renderTagBy={[Function]}
          tags={
            Array [
              Object {
                "selected": false,
                "tag": "one",
                "text": "one",
              },
            ]
          }
        >
          <div
            className="flex flex--wrap undefined"
          >
            <Button
              className="btn radius-top-left tag-one-inverted-coloring"
              data-testid="tag-btn-one"
              key="tag-filter-one-0"
              onClick={[Function]}
              type="button"
            >
              <button
                className="btn radius-top-left tag-one-inverted-coloring"
                data-testid="tag-btn-one"
                onClick={[Function]}
                type="button"
              >
                one
              </button>
            </Button>
          </div>
        </InlineTagList>
      </div>
    </div>
  </Card>
</TagFilter>
`;

exports[`Testing the TagFilter component Should toggle all selection values 1`] = `
<TagFilter
  onSelectionChange={[MockFunction]}
  tags={
    Array [
      Object {
        "selected": false,
        "tag": "one",
        "text": "one",
      },
      Object {
        "selected": false,
        "tag": "two",
        "text": "two",
      },
    ]
  }
>
  <Card
    title="Tags"
  >
    <div
      className="container high-contrast-coloring margin-bottom radius "
    >
      <div
        className="flex flex--column"
        data-testid="titleDescContainer"
      >
        <h6>
          Tags
        </h6>
      </div>
      <div
        className="flex"
      >
        <div>
          <Button
            className="btn btn--inverted margin-right"
            data-testid="allBtn"
            onClick={[Function]}
            type="button"
          >
            <button
              className="btn btn--inverted margin-right"
              data-testid="allBtn"
              onClick={[Function]}
              type="button"
            >
              All
            </button>
          </Button>
        </div>
        <InlineTagList
          onTagClick={[Function]}
          renderTagBy={[Function]}
          tags={
            Array [
              Object {
                "selected": false,
                "tag": "one",
                "text": "one",
              },
              Object {
                "selected": false,
                "tag": "two",
                "text": "two",
              },
            ]
          }
        >
          <div
            className="flex flex--wrap undefined"
          >
            <Button
              className="btn radius-top-left tag-one-inverted-coloring"
              data-testid="tag-btn-one"
              key="tag-filter-one-0"
              onClick={[Function]}
              type="button"
            >
              <button
                className="btn radius-top-left tag-one-inverted-coloring"
                data-testid="tag-btn-one"
                onClick={[Function]}
                type="button"
              >
                one
              </button>
            </Button>
            <Button
              className="btn radius-bottom-right tag-two-inverted-coloring"
              data-testid="tag-btn-two"
              key="tag-filter-two-1"
              onClick={[Function]}
              type="button"
            >
              <button
                className="btn radius-bottom-right tag-two-inverted-coloring"
                data-testid="tag-btn-two"
                onClick={[Function]}
                type="button"
              >
                two
              </button>
            </Button>
          </div>
        </InlineTagList>
      </div>
    </div>
  </Card>
</TagFilter>
`;

exports[`Testing the TagFilter component Should toggle all selection values 2`] = `
<TagFilter
  onSelectionChange={
    [MockFunction] {
      "calls": Array [
        Array [
          Array [
            Object {
              "selected": true,
              "tag": "one",
              "text": "one",
            },
            Object {
              "selected": true,
              "tag": "two",
              "text": "two",
            },
          ],
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  tags={
    Array [
      Object {
        "selected": true,
        "tag": "one",
        "text": "one",
      },
      Object {
        "selected": true,
        "tag": "two",
        "text": "two",
      },
    ]
  }
>
  <Card
    title="Tags"
  >
    <div
      className="container high-contrast-coloring margin-bottom radius "
    >
      <div
        className="flex flex--column"
        data-testid="titleDescContainer"
      >
        <h6>
          Tags
        </h6>
      </div>
      <div
        className="flex"
      >
        <div>
          <Button
            className="btn btn--inverted margin-right"
            data-testid="allBtn"
            onClick={[Function]}
            type="button"
          >
            <button
              className="btn btn--inverted margin-right"
              data-testid="allBtn"
              onClick={[Function]}
              type="button"
            >
              All
            </button>
          </Button>
        </div>
        <InlineTagList
          onTagClick={[Function]}
          renderTagBy={[Function]}
          tags={
            Array [
              Object {
                "selected": true,
                "tag": "one",
                "text": "one",
              },
              Object {
                "selected": true,
                "tag": "two",
                "text": "two",
              },
            ]
          }
        >
          <div
            className="flex flex--wrap undefined"
          >
            <Button
              className="btn radius-top-left tag-one-inverted-coloring"
              data-testid="tag-btn-one"
              key="tag-filter-one-0"
              onClick={[Function]}
              type="button"
            >
              <button
                className="btn radius-top-left tag-one-inverted-coloring"
                data-testid="tag-btn-one"
                onClick={[Function]}
                type="button"
              >
                one
              </button>
            </Button>
            <Button
              className="btn radius-bottom-right tag-two-inverted-coloring"
              data-testid="tag-btn-two"
              key="tag-filter-two-1"
              onClick={[Function]}
              type="button"
            >
              <button
                className="btn radius-bottom-right tag-two-inverted-coloring"
                data-testid="tag-btn-two"
                onClick={[Function]}
                type="button"
              >
                two
              </button>
            </Button>
          </div>
        </InlineTagList>
      </div>
    </div>
  </Card>
</TagFilter>
`;
